base/abstractarray.jl

```diff
@@ -252,15 +267,20 @@ end
 
 # this method must be separate from the above since src might not have a length
 function copy!(dest::AbstractArray, doffs::Integer, src, soffs::Integer, n::Integer)
-    n < 0 && throw(BoundsError())
+    n < 0 && throw(BoundsError(dest, n))
     n == 0 && return dest
     dmax = doffs + n - 1
     if (dmax > length(dest)) | (doffs < 1) | (soffs < 1)
-        throw(BoundsError())
+        dmax > length(dest) && throw(BoundsError(dest, dmax))
+        doffs < 1 && throw(BoundsError(dest, doffs))
+        throw(BoundsError(src, soffs))
```

Comment:
Your tests will be a little more sane if this also throws an `ArgumentError`.  But I'm not convinced that this is the way to go.